examples/lambda/wcbv-reasonable/PrelimsScript.sml

```diff
@@ -0,0 +1,491 @@
+   for inspiration
+*)
+open HolKernel Parse boolLib bossLib;
+open arithmeticTheory;
+open listTheory;
+
+val _ = new_theory "Prelims";
+
+(* ------------------
+	 Natural numbers
+   ------------------ *)
+
+Theorem size_induction:
+	∀f p. (∀x. ((∀y. f y < f x ⇒ p y) ⇒ p x)) ⇒ (∀x. p x)
+Proof
+	ntac 4 strip_tac >>
+	`(∀y. f y < f x ⇒ p y)` suffices_by gs[] >>
+	`∀n y. f y < n ⇒ p y` suffices_by metis_tac[] >>
+	Induct_on `n` >> rw[]
+QED
+
+(* ------------------
+	   	 Lists
+   ------------------ *)
+
+Definition nth_error:
+	nth_error 0 (h::_) = SOME h ∧
+	nth_error (SUC n) (_::t) = nth_error n t ∧
+    nth_error _ _ = NONE
+End
+
+Theorem nth_error_lt_Some:
+	∀n H. n < LENGTH H ⇒ ∃x. nth_error n H = SOME x
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1]
+	>- (qexists_tac `EL 0 H` >> Induct_on `H` >> rw[nth_error])
+    >> Induct_on `H` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_Some_lt:
+	∀n H x. nth_error n H = SOME x ⇒ n < LENGTH H
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_map:
+	∀n H a f. nth_error n (MAP f H) = SOME a ⇒ ∃b. nth_error n H = SOME b ∧ a = f b
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem map_nth_error:
+	∀n H x f. nth_error n H = SOME x ⇒ nth_error n (MAP f H) = SOME (f x)
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem nth_error_NONE_lt:
+	∀n H. nth_error n H = NONE ⇒ LENGTH H ≤ n
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem nth_error_lt_NONE:
+	∀n H. LENGTH H ≤ n ⇒ nth_error n H = NONE
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `H` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_SOME_lemma:
+	∀n H h t x.
+		nth_error n (h::t) = SOME x ⇒
+		1 <= n ⇒
+		nth_error (n-1) t = SOME x
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `H` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_SOME_in_H:
+	∀n H x. nth_error n H = SOME x ⇒ MEM x H
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem nth_error_In:
+  ∀l n x.
+    nth_error n l = SOME x ⇒ MEM x l
+Proof
+  metis_tac[nth_error_SOME_in_H]
+QED
+
+Theorem nth_error_app1:
+	∀l l' n.
+		n < LENGTH l ⇒
+    	nth_error n (l++l') = nth_error n l
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `l` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_app2:
+	∀l l' n.
+		LENGTH l ≤ n ⇒
+    	nth_error n (l++l') = nth_error (n-LENGTH l) l'
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `l` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Inductive Forall:
```

Comment:
Please rewrite to use ``EVERY`` instead, which this is a copy of.

Reply:
Done - removed `Forall` and all the related lemmas since they're not used in the later proofs.